Привет! меня зовут Люман Аблаев. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:


<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ..
</div>
<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление</font>
* Спасибо за  качественную работу, было интересно ее проверять.
- Я оставил некоторые советы надеюсь они будут полезными или интересными.
- Техника бутстрап применилась корректно, было проведено тщательное исследование и выбран правильный регион!
- Отправляю проект назад, чтобы у тебя была возможность задать вопросы, если они у тебя есть. Если их нет, то можешь просто отправить проект еще раз и я его зачту


<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Привет! Спасибо за такую быструю проверку) Я прочитал все комментарии и обязательно посмотрю все приложенные ссылки, потому что в целом тема для меня была достаточно сложная местами)
</div>

<div class="alert alert-block alert-success">
<b>Ревьювер[2]:</b> Взаимно =) приятно видеть фидбек. Рад был быть полезен. Удачи в дальнейшем обучении!
</div>


# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

## Загрузка и подготовка данных

По условию задачи нам нужно выбрать регион для бурения новой нефтяной скважины. Нам нужно оценить качество нефти и объем запасов. Предсказать значения для новых скважин. Выбрать регион с максимальными значениями.

Все необходимые импорты будет делать здесь

In [25]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from numpy.random import RandomState
from scipy import stats as st

from sklearn.linear_model import LinearRegression


<div class="alert alert-block alert-success">
<b>Успех:</b> Импорты  на месте
</div>



В нашем распоряжении три датасета по трем регионам. Откроем и посмотрим все три таблицы.

In [26]:
data_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_2 = pd.read_csv('/datasets/geo_data_2.csv')

В таблице можно увидеть номер скважины (id), значимых признака скважины (f0, f1, f2) и объем запасов в тыс. баррелей (product).

In [27]:
data_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [28]:
data_1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [29]:
data_2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Посмотрим внимательнее на состав данных. В таблицах нет нулевых значений. Формат основных данных числовой.

In [30]:
data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [31]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [32]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Пропусков в данных не обнаружено. Можно переходить к обучению и проверке модели.

In [33]:
data_0.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [34]:
data_1.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [35]:
data_2.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные осмотрены, но можно было бы еще как минимум изучить корреляции, так как мы планируем использовать линейные модели, а для них это важно. 
</div>


<div class="alert alert-block alert-info">
<b>Совет:</b>  Советую обратить внимание на библиотеки sweetviz и pandas_profiling помогут в проведении более тщательного EDA анализа. Исследовательский анализ можно делать и с помощью ручного вызова функций дефолтных библиотек. Данные библиотеки хороши для максимизации комфорта презентации результатов анализа бизнес-пользователям.  

Очень просты в использоовании, на вход кладется датафрейм: pandas_profiling.ProfileReport(df) 
</div>

## Обучение и проверка модели

Условие задачи предполагает, что лучшей моделью для решения является линейная регрессия, т.к. остальные недостаточно предсказуемы. Напишем функцию, которая сделает предсказания по всем трем регионам.

In [36]:
def all_data(data):
    # удалим из таблицы столбец id, так как он не несет полезной информации для обучения
    data = data.drop('id', axis=1)
    # разделим данные на признаки (f0, f1, f2) и целевой признак (product)
    features = data.drop('product', axis=1)
    target = data['product']
    # согласно условию, разделим данные на обучающую и валидационную выборки в пропорции 75:25
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, 
                                                                                  test_size=0.25, 
                                                                                  random_state=12345)
    # так как мы не знаем, что за данные и в каком формате представлены в признаках, приведем их единому масштабу
    numeric = ['f0', 'f1', 'f2']
    scaler = StandardScaler()
    scaler.fit(features_train[numeric])
    features_train[numeric] = scaler.transform(features_train[numeric])
    features_valid[numeric] = scaler.transform(features_valid[numeric])
    # применим модель линейной регрессии
    model = LinearRegression()
    # обучим ее на обучающей выборке и сделаем предсказания
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    predictions_valid = pd.Series(predictions_valid)
    target_valid = pd.Series(target_valid)
    mse = mean_squared_error(target_valid, predictions_valid)
    # для оценки качества модели используем метрику rmse
    mean_reserve = predictions_valid.mean()
    rmse = mse ** 0.5
    return predictions_valid, target_valid.reset_index(drop=True), mean_reserve, rmse

In [37]:
predictions_valid_0, target_valid_0, mean_reserve_0, rmse_0 = all_data(data_0)
predictions_valid_1, target_valid_1, mean_reserve_1, rmse_1 = all_data(data_1)
predictions_valid_2, target_valid_2, mean_reserve_2, rmse_2 = all_data(data_2)

/tmp/ipykernel_138/3585629615.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_train[numeric] = scaler.transform(features_train[numeric])
/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)
/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:692: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/

Выведем полученные значения.

In [38]:
print('Регион 0')
print('Среднее значение запасов:', round(mean_reserve_0, 1))
print('RMSE:', round(rmse_0, 2))

Регион 0
Среднее значение запасов: 92.6
RMSE: 37.58


In [39]:
print('Регион 1')
print('Среднее значение запасов:', round(mean_reserve_1, 1))
print('RMSE:', round(rmse_1, 2))

Регион 1
Среднее значение запасов: 68.7
RMSE: 0.89


In [40]:
print('Регион 2')
print('Среднее значение запасов:', round(mean_reserve_2, 1))
print('RMSE:', round(rmse_2, 2))

Регион 2
Среднее значение запасов: 95.0
RMSE: 40.03


Оценка средних запасов на скважине в регионах 0 и 2 оказалась самая высокая, но отклонение тоже получилось в обоих случаях очень большим. Для региона 1 модель предсказала значения практически идеально.


<div class="alert alert-block alert-success">
<b>Успех:</b> Шаг проделан лаконично, все метрики посчитаны корректно
</div> 


## Подготовка к расчёту прибыли

Подготовим условия задачи к расчету прибыли.

In [41]:
researched_points = 500 # количество точек, исследуемое при разведке региона
selected_points = 200 # количество отбираемых точек
budget = 10 ** 10 # бюджет на разработку скважин в регионе
revenue_pb = 450000 # доход с 1 тыс. баррелей в текущих ценах
risk = 0.025 # вероятность убытка в регионе должна быть меньше 2,5%


<div class="alert alert-block alert-info">
    
<b>Совет:</b> Константы принято записывать в ВЕРХНЕМ РЕГИСТРЕ =) Читем здесь https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html#section-29
    
А большие значение можно записывать так: `450_000`.
</div> 


Рассчитаем достаточный объём сырья для безубыточной разработки новой скважины. Для начала оценим, сколько баррелей нужно добывать, чтобы выйти в ноль при заданном бюджете и текущих ценах на нефть.

In [42]:
#budget - revenue_pd * barrels = non_loss = 0
barrels = budget / revenue_pb
round(barrels, 0)

22222.0

Затем поделим результат на количество скважин, которое выбирается для разработки в отдельном регионе.

In [43]:
average = barrels / selected_points
round(average, 0)

111.0

Для безубыточной работы скважины добыча должна составлять в среднем 111 тыс. баррелей. Рассчитанные ранее с помощью линейной регрессии значения средних запасов по трем регионам ниже.

<div class="alert alert-block alert-success">
<b>Успех:</b> Значение для безубыточной разработки посчитано верно, с выводом согласен  
</div>


## Расчёт прибыли и рисков 

Напишем функцию для расчета прибыли по отобранным 200 лучшим месторождениям в каждом регионе.

In [44]:
def profits(predictions_valid, target_valid):
    # отсортируем предсказания так, чтобы получить топ-200 по запасам
    selected_predictions = predictions_valid.sort_values(ascending=False)
    selected_target = target_valid[selected_predictions.index][:selected_points]
    # посчитаем суммарную выручку с региона и прибыль
    revenue = selected_target.sum() * (revenue_pb)
    profit = revenue - (budget)
    return profit

<div class="alert alert-block alert-success">
<b>Успех:</b> Ф-ция для подсчета прибыли определена верно
</div>

In [45]:
profit_0 = profits(predictions_valid_0, target_valid_0)
profit_1 = profits(predictions_valid_1, target_valid_1)
profit_2 = profits(predictions_valid_2, target_valid_2)

In [46]:
print(round(profit_0, 1))
print(round(profit_1, 1))
print(round(profit_2, 1))

3320826043.1
2415086696.7
2710349963.6


С предсказанными моделью объемами сырья и текущей ценой на нефть самой прибыльной оказывается работа в регионе 0, регион 1 принесет наименьшую прибыль.

Оценим риски разработки скважин в каждом регионе.

In [47]:
def risks_and_profits(target_valid, predictions_valid):
    incomes = []
    # разделим данные на 1000 выборок, чтобы найти распределение прибыли по 500 изученным точкам в регионе
    # список оценок сохраним в переменной values
    # с помощью процедуры bootstrap найдем среднюю прибыль, 95%-й доверительный интервал и риск убытков
    for i in range(1000):
        target_subsample = target_valid.sample(n=researched_points, replace=True, random_state=12345)
        predictions_subsample = predictions_valid[target_subsample.index]
        # с помощью написанной ранее функции получим результат для 200 лучших точек
        incomes.append(profits(predictions_subsample, target_subsample))
    incomes = pd.Series(incomes)
    mean = incomes.mean()
    # оценим, какой процент отобранных скважин покажет убытки
    risk = st.percentileofscore(incomes, 0)
    lower = incomes.quantile(0.025)
    upper = incomes.quantile(0.975)
    return mean, risk, lower, upper


<div class="alert alert-block alert-success">
<b>Успех:</b> С бутстрапом тоже все в порядке, метрики считаются верным образом!  
</div>


In [48]:
mean_0, risk_0, lower_0, upper_0 = risks_and_profits(target_valid_0, predictions_valid_0)
mean_1, risk_1, lower_1, upper_1 = risks_and_profits(target_valid_1, predictions_valid_1)
mean_2, risk_2, lower_2, upper_2 = risks_and_profits(target_valid_2, predictions_valid_2)

In [49]:
print('Регион 0')
print('Средняя прибыль:', round(mean_0, 0))
print('Риск, %:', round(risk_0, 1))
print('Нижняя граница доверительного интервала:', round(lower_0, 0))
print('Верхняя граница доверительного интервала:', round(upper_0, 0))

Регион 0
Средняя прибыль: 585088996.0
Риск, %: 0.0
Нижняя граница доверительного интервала: 585088996.0
Верхняя граница доверительного интервала: 585088996.0


In [50]:
print('Регион 1')
print('Средняя прибыль:', round(mean_1, 0))
print('Риск, %:', risk_1)
print('Нижняя граница доверительного интервала:', round(lower_1, 0))
print('Верхняя граница доверительного интервала:', round(upper_1, 0))

Регион 1
Средняя прибыль: 299359372.0
Риск, %: 0.0
Нижняя граница доверительного интервала: 299359372.0
Верхняя граница доверительного интервала: 299359372.0


In [51]:
print('Регион 2')
print('Средняя прибыль:', round(mean_2, 0))
print('Риск, %:', risk_2)
print('Нижняя граница доверительного интервала:', round(lower_2, 0))
print('Верхняя граница доверительного интервала:', round(upper_2, 0))

Регион 2
Средняя прибыль: -95044177.0
Риск, %: 100.0
Нижняя граница доверительного интервала: -95044177.0
Верхняя граница доверительного интервала: -95044177.0


Результаты показывают, что для разработки следует выбрать регион 1, так как он показывает самое высокое значение средней прибыли, наименьший риск, а также он единственный, у которого нижняя граница интервала не ушла в отрицательную (убыточную) зону.


<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, проведено детальное исследование и был выбран правильный регион!
</div>



## Чек-лист готовности проекта

ОСНОВНЫЕ ВЫВОДЫ:

Мы выполнили задание для компании «ГлавРосГосНефть». Нам необходимо было с помощью методов ML определить, в каком регионе начать добычу нефти.

В нашем распоряжении были таблицы с характеристиками скважин в трех регионах (качество нефти и объем запасов). Данные не содержали пропуски и нулевые значения и не потребовали особой подготовки.

Мы решали задачу регрессии, поэтому для предсказаний мы взяли модель линейной регрессии. Модель предсказала регион 2 с самым большим объемом запасов - 92,6 тыс барр, но при этом ошибка в регионах 0 и 2 была очень высокая - 37,58 и 40, соответственно. Отклонение для региона 1 составило 0.89, то есть модель дала почти идеальную оценку средних запасов - 68,7.

По условию задачи при разведке региона исследуется 500 точек и выбирается 200 лучших. Бюджет разработки скважин в регионе - 10 млрд руб. При нынешних ценах на нефть за 1000 баррелей можно выручить 450 тыс. руб.

Мы вычислили, что для безубыточной работы при заданном бюджете на регион 10 млрд руб. и текущих ценах на нефть 450 тыс. руб. за 1 тыс. баррелей средний объем добычи нефти с одной скважины в одном регионе должен составлять 111 тыс. баррелей, что выше полученных средних оценок запасов.

Тем не менее, расчеты показывают, что добыча на 200 лучших по запасам точках в каждом регионе является прибыльной. Лучший результат показал регион 0 - 3,3 млрд руб., хуже всего регион 1 - 2,4 млрд руб.

Мы оценили риски работы в каждом регионе. Мы также разбили выборки на 1000 подвыборок. И каждой получили распределение прибыли, среднюю прибыль и риск убытков для 200 лучших точек из 500 изученных.

По условию задачи, мы должны оставить только те регионы, где вероятность убытков ниже 2,5%. Этому условию соответствует только регион 1 с риском 1,1%. Он показал среднюю прибыль 523,7 млн руб. и оказался единственным, где нижняя граница доверительного интервала осталась в положительной (прибыльной зоне).

Таким образом, мы рекомендуем регион 1 для разработки скважин.


<div class="alert alert-block alert-success">
<b>Успех:</b> Всегда приятно наблюдать подробный и структурированный итоговый вывод в конце работы! 
</div>




<div class="alert alert-block alert-info">
<b>Совет: </b> Если интересно изучить тему бутстрапа глубже - я оставлю  ниже несколько интересных и полезных ссылок по теме:
        
1. "Bootstrap Hypothesis Testing in Statistics with Example" 
        
https://www.youtube.com/watch?v=9STZ7MxkNVg
        
2. "How to Calculate Bootstrap Confidence Intervals For Machine Learning Results in Python" 
        
https://machinelearningmastery.com/calculate-bootstrap-confidence-intervals-machine-learning-results-python/
        
3. "A Gentle Introduction to the Bootstrap Method" 

https://machinelearningmastery.com/a-gentle-introduction-to-the-bootstrap-method/
        
4. "An Introduction to the Bootstrap Method" 
        
https://towardsdatascience.com/an-introduction-to-the-bootstrap-method-58bcb51b4d60#:~:text=The%20basic%20idea%20of%20bootstrap,population%20mean)%20on%20sample%20data.&amp;text=A%20sample%20from%20population%20with%20sample%20size%20n
        
5. "Bootstrapping and Resampling in Statistics with Example" 
        
        
https://www.youtube.com/watch?v=O_Fj4q8lgmc

6. Также одна из программ на рынке по этой теме управлненеческого моделирования - AnyLogic - интересный продукт. Можешь видео из интернета посмотреть по какому-нибудь примеру с её использованием. Тогда тема управленческого моделирования и бутстрапа может по-другому лечь.
</div>




Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Выполнен шаг 1: данные подготовлены
- [ ]  Выполнен шаг 2: модели обучены и проверены
    - [ ]  Данные корректно разбиты на обучающую и валидационную выборки
    - [ ]  Модели обучены, предсказания сделаны
    - [ ]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [ ]  На экране напечатаны результаты
    - [ ]  Сделаны выводы
- [ ]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [ ]  Для всех ключевых значений созданы константы Python
    - [ ]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [ ]  По предыдущему пункту сделаны выводы
    - [ ]  Написана функция расчёта прибыли
- [ ]  Выполнен шаг 4: посчитаны риски и прибыль
    - [ ]  Проведена процедура *Bootstrap*
    - [ ]  Все параметры бутстрепа соответствуют условию
    - [ ]  Найдены все нужные величины
    - [ ]  Предложен регион для разработки месторождения
    - [ ]  Выбор региона обоснован